### ISSUE TITLE
Clarification on behaviour with async processing

### ISSUE THREAD

**Issue by ronkot** [FormidableLabs/pino-lambda#32]
First of all, thanks for the great library!

I was trying out pino-lambda and I'm a bit puzzled about some aspects. It seems that pino-lambda doesn't track AWS request ids same way as Lambdas does when using plain `console.log()` statements. Lambdas seem to use some magic (❓) to be able to attach correct request id to log lines even when there're several parallel async operations running. With pino-lambda calling `logger.withRequest(event, context)` seems to set request id _globally_ which doesn't work when pino is called from parallel async processings.

Maybe I've missed or misunderstood something. Could you check if the logic below is correct or not? Thanks! 🙏 

```js
async function handler(event, context) {
  logger.withRequest(event, context)
  
  console.log('console.log first')
  logger.info('logger.info first')

  await sleep(10000)

  console.log('console.log second')
  logger.info('logger.info second')
}
```

When this function is called twice, the log lines would probably be something like

```
# request 1 comes in
request-1-id console.log first
request-1-id logger.info first

# request 1 processing is awaited, request 2 comes in
request-2-id console.log first
request-2-id logger.info first

# request 2 is awaited, request 1 resumes
request-1-id console.log second
request-2-id logger.info second  // <-- this is wrong, reading request id from global setting, should be request-1-id instead

# request 2 resumes
request-2-id console.log second
request-2-id logger.info second
```

**Comment by carbonrobot** [comment]
The example code you give above is a single request within your Lambda function. Lambda functions only process a single event at a time and do not handle another request until the current request is finished.

In the scenario you proposed above, its not possible to have this occur due to how Lambda handles requests.

Lets say that you make a request to your Lambda function. Before that function is complete, you make a second request. Lambda will either choose to 1. wait until the first request is complete or 2. start up a new Lambda instance to handle your request.

In the first scenario where Lambda waits, the second invocation will call `withRequest` and update the global id. In the second scenario where Lambda starts a new instance you will have an entirely new shared memory space where no global id is yet set.

It's important to remember these semantics only apply in single threaded request environments like Lambda (or GCP cloud functions, azure functions) and an application deployed in EC2 or Kubernetes operates differently.

**Comment by ronkot** [comment]
Thanks for a great clarification @carbonrobot! I thought that lambda has similar execution model as Azure functions have (AFAIK), where single instance serves multiple concurrent requests. Also what lead me to wrong direction was that I run my code inside `serverless-offline` emulator which seems to work differently (more Azure functions way) and for that reason  `pino-lambda` worked as I described above.

I just deployed my `pino-lambda` test code to AWS Lambda and can confirm that your explanation of Lambda execution model is correct and using global state is ok.

Big thanks! 🙏
